contributors/design-proposals/scheduling/scheduling-framework.md

```diff
@@ -1,436 +1,626 @@
-transferred to the scheduling framework.  
-In the rest of this section we describe how various plugins are used to enrich
-this basic workflow. This document focuses on in-process plugins.
-Out-of-process plugins are discussed later in a separate doc.
-
-## Communication and statefulness of plugins
-
-The scheduling framework provides a library that plugins can use to pass
-information to other plugins. This library keeps a map from keys of type string
-to opaque pointers of type interface{}. A write operation takes a key and a
-pointer and stores the opaque pointer in the map with the given key. Other
-plugins can provide the key and receive the opaque pointer. Multiple plugins can
-share the state or communicate via this mechanism.  
-The saved state is preserved only during a single scheduling cycle. At the end
-of a scheduling cycle, this map is destructed. So, plugins cannot keep shared
-state across multiple scheduling cycle. They can, however, update the scheduler
-cache via the provided interface of the cache. The cache interface allows
-limited state preservation across multiple scheduling cycle.  
-It is worth noting that plugins are assumed to be **trusted**. Scheduler does
-not prevent one plugin from accessing or modifying another plugin's state.
-
-## Plugin registration
-
-Plugin registration is done by providing an extension point and a function that
-should be called at that extension point. This step will be something like:
+## Scheduling Cycle
 
-```go
-register("pre-filter", plugin.foo)
-```
-
-The details of the function signature will be provided later.
+The main loop of the scheduler is referred to as a "scheduling cycle". Each
+cycle covers the complete process of assigning one pod to a node (or determining
+that the pod cannot be scheduled). Multiple scheduling cycles are started
+serially, but some parts may run concurrently. (See [Concurrency](#concurrency))
 
 ## Extension points
 
-The following picture shows the scheduling cycle of a Pod and the extension
+The following picture shows the scheduling cycle of a pod and the extension
 points that the scheduling framework exposes. In this picture "Filter" is
-equivalent to "Predicate" in scheduler v1 and "Scoring" is equivalent to
-"Priority function". Plugins are go functions. They are registered to be called
-at one of these extension points. They are called by the framework in the same
-order they are registered for each extension point.  
-In the following sections we describe each extension point in the same order
-they are called in a schedule cycle.
+equivalent to "Predicate" and "Scoring" is equivalent to "Priority function".
+Plugins are registered to be called at one or more of these extension points. In
+the following sections we describe each extension point in the same order they
+are called in a scheduling cycle.
+
+One plugin may register at multiple extension points to perform more complex or
+stateful tasks.
```

Comment:
Autoscaling needs to be able to get a simple yes/no answer for a question 'will pod X fit on node Y'. As such it will generally not work with stateful plugins. We should state that explicitly in the design.